README.md
# tt-bandwidth-manager-gui
This package installs **Traffic Cop** to help you manage bandwidth usage by app or process.

![Traffic Cop](data/traffic-cop.png)

### Limitations
Traffic Cop relies on **[nethogs](https://github.com/raboof/nethogs)** to track network usage. As of version 0.8.5-2, **nethogs** does not track udp packets, which are used by many VoIP apps, including **Zoom** and **Skype**, and thus **Traffic Cop** also doesn't show reliable upload and download rates for udp traffic. However, the backend app, **tt**, *does* properly manage this udp traffic. You can confirm it for yourself if you have a pay-per-MB internet plan and check the counter provided by your ISP periodically during a call to verify that the rate you've set is being properly applied.

### More information
See the README for the **tt-bandwidth-manager** package found at https://github.com/wasta-linux/tt-bandwidth-manager for configuration explanation and other information.

*This app uses an icon based on the "traffic-police" icon created by [Freepik](https://www.flaticon.com/authors/freepik) and found at https://www.flaticon.com/.*


debian/changelog
tt-bandwidth-manager-gui (0.3.4) focal; urgency=medium

  * Fix missing variables when 'systemctl show' command returns an error.

 -- Nate Marti <nate_marti@sil.org>  Tue, 12 Jan 2021 12:46:08 +0100

tt-bandwidth-manager-gui (0.3.3) focal; urgency=medium

  * Add explicit depends (needed for armbian focal).

 -- Nate Marti <nate_marti@sil.org>  Tue, 17 Nov 2020 17:27:55 +0100

tt-bandwidth-manager-gui (0.3.2) focal; urgency=medium

  * Properly handle --version check from commandline.

 -- Nate Marti <nate_marti@sil.org>  Mon, 09 Nov 2020 17:18:55 +0100

tt-bandwidth-manager-gui (0.3.1) focal; urgency=medium

  * Use idle_add when updating Gtk elements from thread.

 -- Nate Marti <nate_marti@sil.org>  Wed, 28 Oct 2020 16:16:13 +0100

tt-bandwidth-manager-gui (0.3.0) focal; urgency=medium

  * Added bandwidth usage fields in the config display.
  * Reduced CPU usage.
  * Improved handling of device switching.

 -- Nate Marti <nate_marti@sil.org>  Wed, 28 Oct 2020 10:58:10 +0100

tt-bandwidth-manager-gui (0.2.2) focal; urgency=medium

  * Fixed bad timestamp comparison causing service crash.

 -- Nate Marti <nate_marti@sil.org>  Fri, 16 Oct 2020 22:00:55 +0100

tt-bandwidth-manager-gui (0.2.1) focal; urgency=medium

  * Fixed variable error when starting app with service stopped.

 -- Nate Marti <nate_marti@sil.org>  Tue, 13 Oct 2020 16:27:27 +0100

tt-bandwidth-manager-gui (0.2.0) focal; urgency=medium

  * Added config display.

 -- Nate Marti <nate_marti@sil.org>  Tue, 13 Oct 2020 07:46:00 +0100

tt-bandwidth-manager-gui (0.1.1) focal; urgency=medium

  * Ensure elevated privileges from CLI.

 -- Nate Marti <nate_marti@sil.org>  Fri, 09 Oct 2020 17:40:36 +0100

tt-bandwidth-manager-gui (0.1) focal; urgency=medium

  * Initial release.

 -- Nate Marti <nate_marti@sil.org>  Thu, 08 Oct 2020 18:33:05 +0100


tests/unit/test.py
#import gi
import unittest

#gi.require_version("Gtk", "3.0")
#from gi.repository import Gtk

from pathlib import Path

from trafficcop import app
from trafficcop import config
from trafficcop import utils

# Assert*() methods here:
# https://docs.python.org/3/library/unittest.html?highlight=pytest#unittest.TestCase

class All(unittest.TestCase):
    def setUp(self):
        self.tests_dir = Path(__file__).parents[1]
        self.pkg_dir = Path(__file__).parents[2]
        self.example_config = self.tests_dir / 'data' / 'tt-example.yaml'

    def test_IGNORE(self):
        # This is for testing new functions.
        pass

    def test_check_diff(self):
        example = self.example_config
        default = self.tests_dir / 'data' / 'tt-default-config.yaml'
        diff = utils.check_diff(example, default)
        self.assertNotEqual(diff, 0)

    def test_ensure_config_backup(self):
        # This only tests that tests/data/tt-config.yaml == tests/data/tt-config.yaml.bak.
        #   It doesn't test the creation of a properly-named new backup file.
        current = self.tests_dir / 'data' / 'tt-config.yaml'
        result = utils.mute(utils.ensure_config_backup, current)
        self.assertTrue(result)

    def tearDown(self):
        pass

class Timestamps(unittest.TestCase):
    def setUp(self):
        self.convert = utils.convert_human_to_epoch

    def test_a_older(self):
        epochA = self.convert("Sat Oct 17 05:32:38 2020")
        epochB = self.convert("Sun Oct 18 05:32:38 2020")
        self.assertGreater(epochB, epochA)

    def test_b_older(self):
        epochA = self.convert("Sat Oct 17 05:32:38 2020")
        epochB = self.convert("Thu Sep 17 05:32:38 2020")
        self.assertGreater(epochA, epochB)

    def test_bad(self):
        epoch = utils.mute(self.convert, "Lun Sep 17 00:00:00 2020")
        self.assertEqual(epoch, "")

    def test_empty(self):
        epoch = self.convert("")
        self.assertEqual(epoch, "")

    def tearDown(self):
        pass

class TtInfo(unittest.TestCase):
    def setUp(self):
        pass

    def test_tt_off(self):
        info = utils.get_tt_info(exe='/dev/null')
        self.assertEqual(info[0], -1)

    def test_tt_on(self):
        info = utils.get_tt_info(exe='/sbin/init')
        self.assertEqual(info[0], 1)

    def test_wait_for_tt_fail(self):
        info = utils.wait_for_tt_start(exe='/dev/null', max=5)
        self.assertEqual(info[0], -1)

    def test_wait_for_tt_pass(self):
        info = utils.wait_for_tt_start(exe='/sbin/init', max=5)
        self.assertEqual(info[0], 1)

    def tearDown(self):
        pass


tests/unit/test_new.py
import unittest
#from pathlib import Path

from trafficcop import app
from trafficcop import config
from trafficcop import handler
from trafficcop import utils
from trafficcop import worker

# Assert*() methods here:
# https://docs.python.org/3/library/unittest.html?highlight=pytest#unittest.TestCase


class New(unittest.TestCase):
    def setUp(self):
        pass

    def test_new(self):
        result = True
        self.assertTrue(result)

    def tearDown(self):
        pass


tests/unit/test_units.py
#import gi
import unittest

#gi.require_version("Gtk", "3.0")
#from gi.repository import Gtk

from pathlib import Path

from trafficcop import app
from trafficcop import config
from trafficcop import utils

# Assert*() methods here:
# https://docs.python.org/3/library/unittest.html?highlight=pytest#unittest.TestCase

class All(unittest.TestCase):
    def setUp(self):
        self.tests_dir = Path(__file__).parents[1]
        self.pkg_dir = Path(__file__).parents[2]
        self.example_config = self.tests_dir / 'data' / 'tt-example.yaml'

    def test_check_diff(self):
        example = self.example_config
        default = self.tests_dir / 'data' / 'tt-default-config.yaml'
        diff = utils.check_diff(example, default)
        self.assertNotEqual(diff, 0)

    def test_ensure_config_backup(self):
        # This only tests that tests/data/tt-config.yaml == tests/data/tt-config.yaml.bak.
        #   It doesn't test the creation of a properly-named new backup file.
        current = self.tests_dir / 'data' / 'tt-config.yaml'
        result = utils.mute(
            utils.ensure_config_backup,
            current
        )
        self.assertTrue(result)

    def test_IGNORE(self):
        # This is for testing new functions.
        pass

    def test_INCOMPLETE_restart_service(self):
        pass
        #app.app.do_startup()
        # Get widget state.
        #timestamp_pre = app.app.label_applied.get_text()
        #app.app.restart_service()
        #timestamp_post = app.app.label_applied.get_text()
        #self.assertNotEqual(timestamp_pre, timestamp_post)

    def tearDown(self):
        pass


trafficcop/utils.py
import contextlib
import locale
import netifaces
import os
import psutil
import re
import shutil
import subprocess
import time


@contextlib.contextmanager
def setlocale(*args, **kw):
    saved = locale.setlocale(locale.LC_ALL)
    yield locale.setlocale(*args, **kw)
    locale.setlocale(locale.LC_ALL, saved)

def mute(func, *args, **kwargs):
    with open(os.devnull, 'w') as devnull:
        with contextlib.redirect_stdout(devnull):
            output = func(*args, **kwargs)
    return output

def get_net_device():
    '''
    Return the gateway internet device.
    If there are multiple gateways found, then the highest priority device is returned:
        1. Bluetooth (AF_BLUETOOTH)
        2. Cellular (AF_PPPOX)
        3. IPv6 (AF_INET6)
        4. IPv4 (AF_INET)
    '''
    gws = netifaces.gateways()
    families = [
        netifaces.AF_BLUETOOTH,
        netifaces.AF_PPPOX,
        netifaces.AF_INET6,
        netifaces.AF_INET,
    ]
    for family in families:
        try:
            device = gws['default'][family][1]
            break
        except KeyError:
            device = None
    return device

def convert_epoch_to_human(epoch):
    human = time.ctime(epoch)
    return human

def convert_human_to_epoch(human):
    if human:
        with setlocale(locale.LC_TIME, "C"):
            try:
                str = time.strptime(human, '%a %b %d %H:%M:%S %Y') # Tue Oct 13 05:59:00 2020
                # Convert object to epoch format.
                epoch = time.mktime(str) # Tue 2020-10-13 05:59:00 WAT
            except ValueError as v:
                print(repr(v))
                epoch = ''
            except Exception as e:
                print(repr(e))
                epoch = ''
    else:
        epoch = human
    return epoch

def convert_human_to_log(human):
    # Convert human to object.
    #   Doesn't work: prob b/c my locale is FR but human is reported in EN.
    with setlocale(locale.LC_TIME, "C"):
        str = time.strptime(human, '%a %b %d %H:%M:%S %Y') # Tue Oct 13 05:59:00 2020
        # Convert object to log format.
        log = time.strftime('%a %Y-%m-%d %H:%M%S %Z', str) # Tue 2020-10-13 05:59:00 WAT
        return log

def convert_bytes_to_human(bytes_per_sec):
    # "human" means "3 significant digits, changing power as necessary."
    float = bytes_per_sec
    unit = 'B/s'
    if len(str(int(float))) > 3:
        # Switch to KB.
        float = float / 1000
        unit = 'KB/s'
        if len(str(int(float))) > 3:
            # Switch to MB.
            float = float / 1000
            unit = 'MB/s'
            if len(str(int(float))) > 3:
                # Switch to GB.
                float = float / 1000
                unit = 'GB/s'
    return [float, unit]

def get_tt_info(exe='/usr/bin/tt'):
    procs = psutil.process_iter(attrs=['pid', 'cmdline', 'create_time'])
    for proc in procs:
        try:
            if exe == proc.cmdline()[1]:
                proc.dev = proc.cmdline()[2]
                proc.start = convert_epoch_to_human(proc.create_time())
                return proc.pid, proc.start, proc.dev
        except (psutil.NoSuchProcess, psutil.AccessDenied, psutil.ZombieProcess, IndexError):
            pass
    return -1, '', ''

def get_file_mtime(file):
    statinfo = os.stat(file)
    mtime = convert_epoch_to_human(statinfo.st_mtime)
    return mtime

def wait_for_tt_start(exe='/usr/bin/tt', max=100):
    '''
    Wait for service status to start, otherwise update_service_props() may not
    get the correct info.
    '''
    ct = 0
    # Initially assume that tt is not running.
    tt_pid, tt_start, tt_dev = -1, '', ''
    while ct < max:
        tt_pid, tt_start, tt_dev = get_tt_info(exe)
        if psutil.pid_exists(tt_pid):
            return tt_pid, tt_start, tt_dev
        time.sleep(0.1)
        ct += 1
    return tt_pid, tt_start, tt_dev

def check_diff(file1, file2):
    result = subprocess.run(
        ["diff", file1, file2],
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL
    )
    return result.returncode

def ensure_config_backup(current):
    '''
    Make a backup of user config; add index to file name if other backup already exists.
    '''
    already = "Current config already backed up at"
    name = current.stem
    suffix = ".yaml.bak"
    backup = current.with_suffix(suffix)
    if not backup.exists():
        shutil.copyfile(current, backup)
        return True
    diff = check_diff(current, backup)
    if diff == 0:
        print(already, backup)
        return True
    # The backup file exists and is different from current config:
    #   need to choose new backup file name and check again.
    # Add index to name.
    i = 1
    # Set new backup file name.
    backup = current.with_name(name + '-' + str(i)).with_suffix(suffix)
    if not backup.exists():
        shutil.copyfile(current, backup)
        return True
    diff = check_diff(current, backup)
    if diff == 0:
        print(already, backup)
        return True
    while backup.exists():
        # Keep trying new indices until an available one is found.
        i += 1
        backup = current.with_name(name + '-' + str(i)).with_suffix(suffix)
        if not backup.exists():
            shutil.copyfile(current, backup)
            return True
        diff = check_diff(current, backup)
        if diff == 0:
            print(already, backup)
            return True

def update_global_scope():
    '''
    Update system bandwidth usage for the Global scope.
    '''
    net_io = psutil.net_io_counters(pernic=True)
    dev = get_net_device()
    if not dev:
        # No current connection.
        return [0, 0]
    bytes_up = net_io[dev].bytes_sent
    bytes_dn = net_io[dev].bytes_recv
    return [bytes_up, bytes_dn]

def update_scopes(scopes, queue, store):
    '''
    Retrieve items from nethogs queue and show updated download and upload rates.
    '''
    # Get time of current iteration.
    epoch = time.time()

    # Move current scopes dict's 'new' entries to 'last'.
    for scope in scopes.keys():
        scopes[scope]['last'] = scopes[scope]['now'].copy()

    # Get fresh proc list.
    proc_list = psutil.process_iter(attrs=['name', 'exe', 'cmdline'])

    # Update scopes dict 'new' entries.
    while not queue.empty():
        line = queue.get().split()
        exe_pid_usr = line[0]
        b_up = int(float(line[-2]))
        b_dn = int(float(line[-1]))
        scope = match_cmdline_to_scope(exe_pid_usr, store, proc_list)
        if not scope:
            # Not matched; will be counted in 'Global'.
            continue
        if scope not in scopes.keys():
            # Initialize scopes[scope].
            scopes[scope] = {
                'last': {
                    'time': None,
                    'bytes_up': None,
                    'bytes_dn': None,
                },
                'now': {}
            }

        # Add epoch time to 'now'.
        scopes[scope]['now']['time'] = epoch
        # Update bytes for current scope.
        scopes[scope]['now']['bytes_up'] = b_up
        scopes[scope]['now']['bytes_dn'] = b_dn

    # Update Global scope.
    bytes = update_global_scope()
    b_up = bytes[0]
    b_dn = bytes[1]
    if 'Global' not in scopes.keys():
        scopes['Global'] = {
                'last': {
                    'time': None,
                    'bytes_up': None,
                    'bytes_dn': None,
                },
                'now': {}
        }
    scopes['Global']['now']['time'] = epoch
    scopes['Global']['now']['bytes_up'] = b_up
    scopes['Global']['now']['bytes_dn'] = b_dn

    return scopes

def match_cmdline_to_scope(exe_pid_usr, store, proc_list):
    # Strip pid and user from cmdline.
    cmdline_list = exe_pid_usr.split('/')
    pid = cmdline_list[-2]
    exe = '/'.join(cmdline_list[:-2])

    # Get scope names, match-type, and match-str from store.
    scopes = {}
    for row in store:
        if row[0] == 'Global':
            continue
        scopes[row[0]] = row[11:]

    # Get cmdlines from proces_iter.
    match_exe_pid_usr_and_proc = {}
    for proc in proc_list:
        try:
            p_pid = str(proc.pid)
        except psutil.NoSuchProcess:
            continue
        if p_pid == pid:
            match_exe_pid_usr_and_proc = proc.info
            break

    # Match cmdline with scope.
    scope = None
    if not match_exe_pid_usr_and_proc:
        return None

    for k, v in scopes.items():
        # k = scope; v = [match-type, match-str]
        if v[0] == 'name':
            match = re.match(v[1], match_exe_pid_usr_and_proc['name'])
            if match:
                scope = k
                break
        elif v[0] == 'exe':
            # See if scope exe matches proc exe.
            match = re.match(v[1], match_exe_pid_usr_and_proc['exe'])
            if match:
                scope = k
                break
        elif v[0] == 'cmdline':
            # See if scope cmdline equals proc cmdline.
            if v[1] == match_exe_pid_usr_and_proc['cmdline']:
                scope = k
                break
        else:
            # Unhandled match-type.
            print('no match for', k, ':', v)
            continue
    return scope

def calculate_data_rates(data):
    elapsed = data['now']['time'] - data['last']['time']
    bytes_up = data['now']['bytes_up'] - data['last']['bytes_up']
    bytes_dn = data['now']['bytes_dn'] - data['last']['bytes_dn']
    rate_up = 0
    rate_dn = 0
    if elapsed > 0:
        rate_up = bytes_up / elapsed
        rate_dn = bytes_dn / elapsed
    return [rate_dn, rate_up]


trafficcop/worker.py
""" Functions that run in background threads. """
# All of these functions run inside of threads and use GLib to communicate back.

import gi
import psutil
import subprocess
import sys
import time

from gi.repository import GLib
from pathlib import Path

from trafficcop import app
from trafficcop import config
from trafficcop import utils


def handle_button_log_clicked():
    # Follow the log since service start time in a terminal window.
    cmd = [
        "gnome-terminal",
        "--geometry=144x24+100+200",
        "--",
        "journalctl",
        "--unit=tt-bandwidth-manager.service",
        "--follow",
        "--output=cat",
        "--no-pager",
        "--since=\'" + app.app.svc_start_time + "\'",
    ]
    if app.app.svc_start_time == 'unknown':
        # Likely due to a kernel/systemd incompatibility.
        cmd.pop() # to remove the "--since" option
    cmd_txt = " ".join(cmd)
    result = subprocess.run(cmd_txt, shell=True)
    print(f"{cmd} -> {result.returncode}")
    return

def handle_button_config_clicked():
    # Open config file in gedit.
    cmd = ["gedit", "/etc/tt-config.yaml"]
    result = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    utils.print_result(cmd, result)

def handle_config_changed():
    pass
    #app.app.update_service_props()
    #read_time = app.app.tt_start
    #print("Service Start Time =", read_time)

def parse_nethogs_to_queue(queue, main_window):
    delay = 1
    device = utils.get_net_device()
    # If no device is given, then all devices are monitored, which double-counts
    #   on gateway device plus tc device.
    cmd = ['nethogs', '-t', '-v2', '-d' + str(delay), device]
    stdout = subprocess.PIPE
    stderr = subprocess.STDOUT
    with subprocess.Popen(cmd, stdout=stdout, stderr=stderr, encoding='utf-8') as p:
        while main_window.is_visible():
            # There is a long wait for each line: sometimes nearly 2 seconds!
            line = p.stdout.readline()
            if line[0] == '/':
                queue.put(line)
            elif line == '' and p.poll() is not None:
                # Process completed. (Shouldn't happen.)
                break

def bw_updater():
    while app.app.window.is_visible():
        time.sleep(1.5)
        # Update the device name.
        GLib.idle_add(app.app.update_device_name)

        # Get all applicable cmdlines & bytes transferred for each scope in config.
        # Sum the total sent for each scope, as well as the total received and give it a timestamp.
        app.app.scopes = utils.update_scopes(app.app.scopes, app.app.net_hogs_q, app.app.config_store)

        # Get the upload and download rates (B/s).
        rates_dict = {}
        for scope, data in app.app.scopes.items():
            if not data['last']['time']:
                continue
            rates = utils.calculate_data_rates(data)

            # Adjust the number to only show 3 digits; change units as necessary (KB/s, MB/s, GB/s).
            human_up = utils.convert_bytes_to_human(rates[0])
            human_dn = utils.convert_bytes_to_human(rates[1])
            rates_dict[scope] = [*human_up, *human_dn]

        # Update the values shown in the treeview.
        GLib.idle_add(config.update_store_rates, app.app.config_store, rates_dict)


trafficcop/app.py
""" Main GUI module. """

import gi
import gzip
import logging
import os
import psutil
import queue
import re
import shutil
import subprocess
import threading
import time

from pathlib import Path
current_file_path = Path(__file__)

gi.require_version("Gtk", "3.0")
from gi.repository import Gio
from gi.repository import GLib
from gi.repository import Gtk

from trafficcop import config
from trafficcop import handler
from trafficcop import utils
from trafficcop import worker


class TrafficCop(Gtk.Application):
    def __init__(self):
        super().__init__(
            application_id='org.wasta.apps.traffic-cop',
            flags=Gio.ApplicationFlags.HANDLES_COMMAND_LINE,
        )
        # Add CLI options.
        self.add_main_option(
            'version', ord('V'), GLib.OptionFlags.NONE, GLib.OptionArg.NONE,
            'Print version number', None
        )

        # Get UI location based on current file location.
        self.ui_dir = '/usr/share/traffic-cop/ui'
        if str(current_file_path.parents[1]) != '/usr/lib/python3/dist-packages':
            self.ui_dir = str(current_file_path.parents[1] / 'data' / 'ui')

        # Define app-wide variables.
        self.tt_pid, self.tt_start, self.tt_dev = utils.get_tt_info()
        self.config_file = Path('/etc/tt-config.yaml')
        self.default_config = Path("/usr/share/tt-bandwidth-manager/tt-default-config.yaml")
        self.config_store = ''
        self.net_hogs_q = queue.Queue()
        self.main_pid = os.getpid()
        self.managed_ports = {}
        self.scopes = {}

    def do_startup(self):
        '''
        do_startup is the setting up of the app, either for "activate" or for "open".
        It runs just after __init__.
        '''
        # Define builder and its widgets.
        Gtk.Application.do_startup(self)

        # Get widgets from glade file, which is defined in __init__.
        self.builder = Gtk.Builder()
        self.builder.add_from_file(self.ui_dir + '/mainwindow.glade')
        self.window = self.builder.get_object('mainwindow')
        self.toggle_active = self.builder.get_object('toggle_active')
        self.toggle_unit_state = self.builder.get_object('toggle_unit_state')
        self.button_restart = self.builder.get_object('button_restart')
        self.label_iface = self.builder.get_object('label_iface')
        self.button_log = self.builder.get_object('button_log')
        self.label_applied = self.builder.get_object('label_applied')
        self.button_apply = self.builder.get_object('button_apply')
        self.button_config = self.builder.get_object('button_config')
        self.button_reset = self.builder.get_object('button_reset')
        self.w_config = self.builder.get_object('w_config')
        self.vp_config = self.builder.get_object('vp_config')

        # Main window.
        self.add_window(self.window)
        self.window.set_icon_name('traffic-cop')

        # Populate config viewport.
        self.treeview_config = self.update_treeview_config()
        self.treeview_config.show()
        self.vp_config.add(self.treeview_config)

        # Populate widget data.
        self.update_info_widgets()
        self.button_apply.set_sensitive(False)

        # Connect GUI signals to Handler class.
        self.builder.connect_signals(handler.Handler())

    def do_command_line(self, command_line):
        '''
        do_command_line runs after do_startup and before do_activate.
        '''
        options = command_line.get_options_dict()
        options = options.end().unpack()

        if not options:
            # No command line args passed: run GUI.
            self.activate()
            return 0

        if 'version' in options:
            # Get version number from debian/changelog.
            if self.runmode == 'uninstalled':
                changelog = Path(__file__).parents[1] / 'debian' / 'changelog'
                with open(changelog) as f:
                    first_line = f.readline()
            else:
                changelog = Path('/usr/share/doc/tt-bandwidth-manager-gui/changelog.gz')
                with gzip.open(changelog) as f:
                    first_line = f.readline().strip().decode()
            # 2nd term in 1st line of changelog; also need to remove parentheses.
            version = first_line.split()[1][1:-1]
            print(f"Traffic Cop (app) / tt-bandwidth-manager-gui (package): {version}")
            exit(0)

    def do_activate(self):
        '''
        do_activate is the displaying of the window. It runs last after do_command_line.
        '''
        # Verify execution with elevated privileges.
        if os.geteuid() != 0:
            self.args.insert(0, 'pkexec')
            subprocess.run(self.args)
            exit()

        # Show window.
        self.window.show()

        # Start tracking operations (self.window must be shown first).
        target = worker.parse_nethogs_to_queue
        args = self.net_hogs_q, self.window
        t_nethogs = threading.Thread(target=target, args=args, name='T-nh')
        t_nethogs.start()

        # Start bandwidth rate updater.
        t_bw_updater = threading.Thread(target=worker.bw_updater, name='T-bw')
        t_bw_updater.start()

        # Verify execution with elevated privileges.
        if os.geteuid() != 0:
            bin = '/usr/bin/traffic-cop'
            print("traffic-cop needs elevated privileges; e.g.:\n\n$ pkexec", bin, "\n$ sudo", bin)
            exit(1)

    def update_service_props(self):
        # Get true service start time.
        self.tt_pid, self.tt_start, self.tt_dev = utils.get_tt_info()

        # Get state of systemd service.
        cmd = [
            "systemctl",
            "show",
            "tt-bandwidth-manager.service",
            "--no-pager",
        ]
        status_output = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        utils.print_result(cmd, status_output)
        if status_output.returncode != 0:
            # Status output error. Probably due to kernel incompatibility after update.
            #   Fall back to trying "systemctl status" command instead.
            self.unit_file_state = 'unknown'
            self.active_state = 'unknown'
            self.svc_start_time = 'unknown'
            cmd.pop(1)
            cmd.insert(1, "status")
            status_output = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
            utils.print_result(cmd, status_output)
            output_list = status_output.stdout.decode().splitlines()
            #print(output_list)
            upat = '\s+Loaded: loaded \(/etc/systemd/system/tt-bandwidth-manager.service; (.*);.*'
            apat = '\s+Active: (.*) since .*'
            for line in output_list:
                try:
                    match = re.match(upat, line)
                    self.unit_file_state = match.group(1)
                except:
                    pass
                try:
                    match = re.match(apat, line)
                    self.active_state = match.group(1).split()[0]
                except:
                    pass

        # Continue with processing of "systemctl show" command output.
        output_list = status_output.stdout.decode().splitlines()

        # Parse output for unit state and active state.
        toggle_unit_state = self.builder.get_object('toggle_unit_state')
        toggle_active = self.builder.get_object('toggle_active')
        upat = '^UnitFileState=.*$'
        apat = '^ActiveState=.*$'
        tpat = '^ExecMainStartTimestamp=.*$'
        for line in output_list:
            try:
                match = re.match(upat, line)
                self.unit_file_state = match.group().split('=')[1]
                continue
            except:
                pass
            try:
                match = re.match(apat, line)
                self.active_state = match.group().split('=')[1]
                continue
            except:
                pass
            try:
                match = re.match(tpat, line)
                # The format for self.tt_start (Tue Oct 13 06:15:14 2020) isn't
                #   compatible with the log file viewer. (Tue 2020-10-13 05:59:00 WAT)
                self.svc_start_time = match.group().split('=')[1]
                continue
            except:
                pass

    def update_state_toggles(self):
        # Update toggle buttons according to current states.
        state = True if self.unit_file_state == 'enabled' else False
        self.toggle_unit_state.set_state(state)
        state = True if self.active_state == 'active' else False
        self.toggle_active.set_state(state)

    def update_device_name(self):
        # Get name of managed interface.
        pid, time, dev = utils.get_tt_info()
        if pid > 0:
            self.label_iface.set_text(dev)
        else:
            self.label_iface.set_text("--")

    def update_config_time(self):
        self.label_applied.set_text(self.tt_start)

    def update_button_states(self):
        # TODO: I need a way to "watch" the config file if setting the "Apply"
        #   button to "sensitive" is ever going to work.
        # Update "Apply" button to be insensitive.
        #self.button_apply.set_sensitive(False)
        # Update "Reset..." button to be insensitive.
        self.button_reset.set_sensitive(False)

        # Set "Apply" button to proper state.
        #config_mtime = utils.get_file_mtime(self.config_file)
        #if self.tt_start and config_mtime > self.tt_start:
            # Update "Apply" button to be sensitive.
            #self.button_apply.set_sensitive(True)

        # Set "Reset..." button to proper state.
        diff_configs = utils.check_diff(self.config_file, self.default_config)
        if not diff_configs == 0:
            # Update "Reset..." button to be sensitive.
            self.button_reset.set_sensitive(True)

    def update_treeview_config(self):
        '''
        This handles both initial config display and updating the display if the
        config file is edited externally.
        '''
        if not self.config_store:
            # App is just starting up; create the store.
            self.config_store = config.convert_yaml_to_store(self.config_file)

        if self.tt_start:
            # Service is running.
            # Check if modified time of config file is newer than last service restart.
            #   The config could have been externally modified. If so, those
            #   changes could be shown here in the app without them actually having
            #   been applied.
            config_mtime = utils.get_file_mtime(self.config_file)
            config_epoch = utils.convert_human_to_epoch(config_mtime)
            tt_epoch = utils.convert_human_to_epoch(self.tt_start)
            if config_epoch > tt_epoch:
                print("WARNING: The config file has been modified since the service started.\nApplying the changes now.")
                self.restart_service()
            else:
                new_config_store = config.convert_yaml_to_store(self.config_file)
                self.config_store = config.update_config_store(self.config_store, new_config_store)

        treeview_config = config.create_config_treeview(self.config_store)
        return treeview_config

    def update_info_widgets(self):
        self.update_service_props()
        self.update_state_toggles()
        self.update_device_name()
        self.update_config_time()
        self.update_button_states()

    def stop_service(self):
        cmd = ["systemctl", "stop", "tt-bandwidth-manager.service"]
        result = subprocess.run(cmd)
        utils.print_result(cmd, result)
        self.update_info_widgets()

    def start_service(self):
        cmd = ["systemctl", "start", "tt-bandwidth-manager.service"]
        result = subprocess.run(cmd)
        utils.print_result(cmd, result)
        self.tt_pid, self.tt_start, self.tt_dev = utils.wait_for_tt_start()
        self.update_info_widgets()
        self.treeview_config = self.update_treeview_config()

    def restart_service(self):
        cmd = ["systemctl", "restart", "tt-bandwidth-manager.service"]
        result = subprocess.run(cmd)
        utils.print_result(cmd, result)
        self.tt_pid, self.tt_start, self.tt_dev = utils.wait_for_tt_start()
        # Check service status and update widgets.
        self.update_info_widgets()
        self.treeview_config = self.update_treeview_config()

    def get_user_confirmation(self):
        text = "The current configuration file will be backed up first."
        label = Gtk.Label(text)
        dialog = Gtk.Dialog(
            'Reset to default configuration?',
            self.window,
            None, #Gtk.Dialog.DESTROY_WITH_PARENT,
            (Gtk.STOCK_CANCEL, Gtk.ResponseType.CANCEL, Gtk.STOCK_OK, Gtk.ResponseType.OK)
        )
        hmarg = 80
        vmarg = 20
        dialog.vbox.set_margin_top(vmarg)
        dialog.vbox.set_margin_bottom(vmarg)
        dialog.vbox.set_margin_start(hmarg)
        dialog.vbox.set_margin_end(hmarg)
        dialog.vbox.set_spacing(20)
        dialog.vbox.pack_start(label, True, True, 5)
        label.show()
        response = dialog.run()
        # CLOSE: -4, OK: -5, CANCEL: -6
        dialog.destroy()
        if response == -5:
            return True
        else:
            return False


app = TrafficCop()
